---
title: "SQANTI-SIM EVALUATION REPORT"
date: "`r paste0('  ','Date', ': ', format(Sys.time(), '%d %B %Y'))`"
output:
  html_document:
    number_sections: false
    toc: true
    toc_float: true
    toc_depth: 3
    theme: spacelab
---

<!-- Knitr settings -->

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, cache = FALSE, echo = FALSE)
```

<!-- CSS settings -->

```{css, echo=FALSE}
.title {
  color: #D31336;
}

h2 {
  color: #3A5A81;
  padding-top: 40px;
}

h4 {
  font-size: 18px;
  font-style: italic;
}
```


<!-- Report generation -->

## Summary {.tabset}

### Overview

```{r}
x <- data.frame(
  "Simulated" = c(res.gene["Total"],
                  nrow(data.index),
                  sum(data.index$structural_category == "FSM"),
                  sum(data.index$structural_category != "FSM")),
  
  "Reconstructed" = c(res.gene["TP"] + res.gene["FP"],
                      nrow(data.class),
                      sum(data.class$structural_category == "FSM"),
                      sum(data.class$structural_category != "FSM"))
)

rownames(x) <- c("Unique genes",
                 "Unique isoforms",
                 "Known transcripts",
                 "Novel transcripts")

datatable(
  x,
  options = list(dom = 't'),
  caption = htmltools::tags$caption(
    style = 'caption-side: bottom; text-align: center;','Table 1: ',
    htmltools::em('Simulated and reconstructed transcriptome summary.'))
  )
```

### Isoform classification

```{r}
x <- data.frame(
  "Simulated" = c(sum(data.index$structural_category == "FSM"),
                  sum(data.index$structural_category == "ISM"),
                  sum(data.index$structural_category == "NIC"),
                  sum(data.index$structural_category == "NNC"),
                  sum(data.index$structural_category == "Genic\nGenomic"),
                  sum(data.index$structural_category == "Antisense"),
                  sum(data.index$structural_category == "Fusion"),
                  sum(data.index$structural_category == "Intergenic"),
                  sum(data.index$structural_category == "Genic\nIntron")),
  
  "Reconstructed" = c(sum(data.class$structural_category == "FSM"),
                  sum(data.class$structural_category == "ISM"),
                  sum(data.class$structural_category == "NIC"),
                  sum(data.class$structural_category == "NNC"),
                  sum(data.class$structural_category == "Genic\nGenomic"),
                  sum(data.class$structural_category == "Antisense"),
                  sum(data.class$structural_category == "Fusion"),
                  sum(data.class$structural_category == "Intergenic"),
                  sum(data.class$structural_category == "Genic\nIntron"))
)

rownames(x) <- c("FSM",
                 "ISM", "NIC", "NNC",
                 "Genic Genomic", "Antisense", "Fusion",
                 "Intergenic", "Genic Intron")

datatable(
  x,
  options = list(dom = 't'),
  caption = htmltools::tags$caption(
    style = 'caption-side: bottom; text-align: center;','Table 2: ',
    htmltools::em('Structural classification of simulated and reconstructed transcripts.'))
  )
```




## Performance metrics {.tabset}

Simulated transcripts were categorized into known simulated transcripts (FSM), which are present in the reference annotation GTF, and novel simulated transcripts that were excluded from the annotation by SQANTI-SIM. The following metrics were computed:

- **Number of simulated isoforms**: Total number of simulated isoforms.
- **True Positive isoforms (TP)**: Isoforms that match junctions with a simulated transcript, and both their TSS and TTS fall within the 50bp threshold.
- **Partial True Positive isoforms (PTP)**: Isoforms that match junctions, but either the TSS or TTS is beyond the threshold.
- **False Negative isoforms (FN)**: Isoforms that were simulated but not detected.
- **False Positive isoforms (FP)**: Transcripts that were detected but were not part of the simulated set.
- **Sensitivity**: TP / Number of simulated transcripts.
- **Precision**: TP / Detected isoforms.
- **F-score**: The harmonic mean of precision and sensitivity.
- **False Discovery Rate**: (FP + PTP) / Detected isoforms.
- **Positive Detection Rate**: (TP + PTP) / (TP + FN).
- **False Detection Rate**: FP / (TP + PTP + FP).

### Tables {.tabset .tabset-dropdown}

#### Full results

```{r}
t1
```

#### Above min. supp. results

```{r}
t2
```

### Plots {.tabset .tabset-dropdown}

#### Compact

```{r}
p2
```

#### Radar chart

```{r, results = 'hide'}
radar.palette <- c("Known"="#750037", "Novel"='#052240', "FSM"="#6BAED6", "ISM"="#FC8D59", "NIC"="#78C679", 
                "NNC"="#EE6A50", "Genic\nGenomic"="#969696", "Antisense"="#66C2A4", "Fusion"="goldenrod1",
                "Intergenic" = "darksalmon", "Genic\nIntron"="#41B6C4")

sqantisim.stats <- rbind(res.full$sqantisim.stats, res.min$sqantisim.stats)[c("Sensitivity", "Precision", "F-score", "Positive_Detection_Rate", "Sensitivity1"),]
rownames(sqantisim.stats) <- c("Sn", "Pr", "F-score", "PDR", "Sn(min.supp.)")
sqantisim.stats <- sqantisim.stats[, colnames(res.full$sqantisim.stats[,res.full$sqantisim.stats["Total",] > 0])]
sqantisim.stats <- sqantisim.stats[c("F-score", "Sn", "Sn(min.supp.)", "Pr", "PDR"), which(!is.na(sqantisim.stats["F-score",]))]
sqantisim.stats.t <- as.data.frame(t(as.matrix(sqantisim.stats)))
sqantisim.stats.t[is.na(sqantisim.stats.t)] <- NA
sqantisim.stats.t <- rbind(1, 0, sqantisim.stats.t)

n <- length(rownames(sqantisim.stats.t)[rownames(sqantisim.stats.t) %in% c("Known", "Novel")])
par(mar=c(1,1,1,1), mfrow=c(1,n))
for (i in 1:n){
  radarchart(
  sqantisim.stats.t[c(1, 2, i+2),], axistype = 1,
  pcol = radar.palette[[rownames(sqantisim.stats.t)[i+2]]], pfcol = scales::alpha(radar.palette[[rownames(sqantisim.stats.t)[i+2]]], 0.5), plwd = 2, plty = 1,
  # Customize the grid
  cglcol = "grey", cglty = 1, cglwd = 0.8,
  # Customize the axis
  axislabcol = "grey", 
  # Variable labels
  vlcex = 0.7, vlabels = colnames(sqantisim.stats.t[c(1, 2, i+2),]),
  caxislabels = c(0, 0.25, 0.5, 0.75, 1), title = rownames(sqantisim.stats.t)[i+2]
)
}

n <- length(rownames(sqantisim.stats.t)[rownames(sqantisim.stats.t) %in% c("1", "2", "Known", "Novel")])
par(mar=c(1,1,1,1), mfrow=c(2,ceiling(n/2)))
for (i in (1+n-2):(nrow(sqantisim.stats.t)-2)){
  radarchart(
  sqantisim.stats.t[c(1, 2, i+2),], axistype = 1,
  pcol = radar.palette[[rownames(sqantisim.stats.t)[i+2]]], pfcol = scales::alpha(radar.palette[[rownames(sqantisim.stats.t)[i+2]]], 0.5), plwd = 2, plty = 1,
  # Customize the grid
  cglcol = "grey", cglty = 1, cglwd = 0.8,
  # Customize the axis
  axislabcol = "grey", 
  # Variable labels
  vlcex = 0.7, vlabels = colnames(sqantisim.stats.t[c(1, 2, i+2),]),
  caxislabels = c(0, 0.25, 0.5, 0.75, 1), title = rownames(sqantisim.stats.t)[i+2]
)
}
```

## Isoform characterization {.tabset}

### Isoform distribution

```{r}
p1
```

### Simulated reads

```{r}
p3
```

### Number of exons

```{r}
p4
```

### Transcript length

```{r}
p5
```

### Isoform complexity

```{r}
p6
```

### Splice junctions

```{r}
p7
```

```{r, results='asis'}
if ('min_cov' %in% colnames(data.index) | 'within_CAGE_peak' %in% colnames(data.index)){
  cat('\n\n## Orthogonal data support {.tabset}\n\n')
}

if ('min_cov' %in% colnames(data.index)){
  cat('\n\n### Short reads\n\n')
  print(p8)
}

if ('within_CAGE_peak' %in% colnames(data.index)){
  cat('\n\n### CAGE peaks\n\n')
  print(p9)
}

if ('min_cov' %in% colnames(data.index) & 'within_CAGE_peak' %in% colnames(data.index)) {
  cat('\n\n### Short reads + CAGE peaks\n\n')
  print(p10)
}
```
